File: terminal-task/utils/observer.js
const { execSync, spawnSync } = require('child_process');
const fs = require('fs');

class Observer {
	constructor() {
		this.service_data = {};
		this.service_history = [];
		this.apps_service_history = {};

		this.text_file = 'x.txt';
		// this.json_file = null;

		this.__to_get_service_data = false;
		this.__to_get_service_history = false;
		this.__to_get_apps_service_history = false;
	}

	set_text_file(text_file) {
		this.text_file = text_file;
	}

	set_command(command) {
		this.command = command;
	}

	to_get_service_data() {
		this.__to_get_service_data = true;
		this.__to_get_service_history = false;
		this.__to_get_apps_service_history = false;
	}

	to_get_service_history() {
		this.__to_get_service_data = false;
		this.__to_get_service_history = true;
		this.__to_get_apps_service_history = false;
	}

	to_get_apps_service_history() {
		this.__to_get_service_data = false;
		this.__to_get_service_history = false;
		this.__to_get_apps_service_history = true;
	}

	connect_json(json_file) {
		this.json_file = json_file;
		// clear this json file
		fs.writeFileSync(this.json_file, JSON.stringify({}, null, 4));
	}

	get_data() {
		if (this.__to_get_service_data) {
			return this.service_data;
		}
		if (this.__to_get_service_history) {
			return this.service_history;
		}
		if (this.__to_get_apps_service_history) {
			return this.apps_service_history;
		}
	}

	run_command(useTempFile = false) {
		if (useTempFile) {
			execSync(this.command + ' > ' + this.text_file);
			return 0;
		}

		let data = '';
		try {
			execSync(`${this.command} > ${this.text_file}`, {
				encoding: 'utf8',
				shell: '/bin/bash',
				cwd: `${__dirname}`,
			});
			data = fs.readFileSync(`${__dirname}/${this.text_file}`, 'utf8');
		} catch (err) {
			if (/not found/.test(err)) {
				throw new Error('Device not found');
			} else {
				throw new Error(err.stderr);
			}
		}

		// console.log(
		// 	data
		// 		.split('\n')
		// 		.filter((el) => el.includes('CONNECT'))
		// 		.join('\n')
		// );
		return data;
	}

	look(save = false, useTempFile = false) {
		this.data = this.run_command(useTempFile);
		// console.log(this.data)
		this.__look__();
		if (save) {
			this.save();
		}
	}

	__look__() {
		// Implement this method in subclasses
	}

	watch(timesec, save = true, useTempFile = false) {
		const current_time = new Date();
		while ((new Date() - current_time) / 1000 < timesec) {
			try {
				this.look(save, useTempFile);
			} catch (e) {
				break;
			}
		}
	}

	save() {
		const dataToSave = this.__to_get_service_data
			? this.service_data
			: this.__to_get_service_history
			? this.service_history
			: this.apps_service_history;

		if (this.json_file) {
			fs.writeFileSync(
				this.json_file,
				JSON.stringify(dataToSave, null, 4)
			);
		}
	}
}

module.exports = Observer;

File: terminal-task/final_test/observer/Observer.py
# Author: Swaroop Dora (AutomateFreely)


from abc import ABC, abstractmethod
import datetime
import subprocess
import json

PACKAGE_NAME = START_TIME = END_TIME = str
class Observer(ABC):
    """
    Observer class is an abstract class that can be inherited to 
    create a new observer of a service where you only need to implement 
    the __look__ and __init__ methods and the rest of the code is taken 
    care of by the Observer class.
    """
    def __init__(self) -> None:
        self.service_data: dict[PACKAGE_NAME, dict] = {}
        self.service_history: list[list[PACKAGE_NAME,
                                        START_TIME, END_TIME]] = []
        self.apps_service_history: dict[PACKAGE_NAME,
                                        list[{START_TIME, END_TIME}]] = {}

        self.text_file: str = "x.txt"
        self.json_file: [str, None] = None

        self.__to_get_service_data = False
        self.__to_get_service_history = False
        self.__to_get_apps_service_history = False

    def set_text_file(self, text_file: str) -> None:
        self.text_file = text_file
    
    def set_command(self, command: str) -> str:
        self.command = command

    def to_get_service_data(self) -> None:
        self.__to_get_service_data = True
        self.__to_get_service_history = False
        self.__to_get_apps_service_history = False

    def to_get_service_history(self) -> None:
        self.__to_get_service_data = False
        self.__to_get_service_history = True
        self.__to_get_apps_service_history = False

    def to_get_apps_service_history(self) -> None:
        self.__to_get_service_data = False
        self.__to_get_service_history = False
        self.__to_get_apps_service_history = True

    def connect_json(self, json_file: str) -> None:
        self.json_file = json_file
        # clear this json file
        with open(self.json_file, "w") as f:
            json.dump({}, f, indent=4)

    def run_command(self) -> None:
        process = subprocess.Popen(
            self.command, stdout=subprocess.PIPE, shell=True)
        process.wait()

    def look(self, save: bool = True) -> None:
        self.run_command()
        self.__look__()
        if save:
            self.save()

    @abstractmethod
    def __look__():
        ...

    def watch(self, timesec: int, save=True) -> None:
        current_time = datetime.datetime.now()
        while (datetime.datetime.now() - current_time).seconds < timesec:
            try:
                self.look()
            except KeyboardInterrupt:
                break

    def save(self) -> None:
        with open(self.json_file, "w") as f:
            if (self.__to_get_service_data):
                json.dump(self.service_data, f, indent=4)
            elif (self.__to_get_service_history):
                json.dump(self.service_history, f, indent=4)
            elif (self.__to_get_apps_service_history):
                json.dump(self.apps_service_history, f, indent=4)
        # if (self.__to_get_service_data):
        #     print(json.dumps(self.service_data))
        # elif (self.__to_get_service_history):
        #     print(json.dumps(self.service_history))
        # elif (self.__to_get_apps_service_history):
        #     print(json.dumps(self.apps_service_history))